### ISSUE TITLE
MIRI fails to detect overflow in signed division

### ISSUE THREAD

**Issue by celinval** [rust-lang/rust#112864]
I tried running MIRI in this code:

```rust
#![feature(core_intrinsics)]
fn main() {
    let ub = unsafe { unchecked_div(i32::MAX, -1) };
    // ---> Uncomment this line to see MIRI correctly finding the overflow.
    // let ub = unsafe { std::intrinsics::unchecked_div(i32::MIN, -1) };
    println!("{ub}");
}

unsafe fn unchecked_div(a: i32, b: i32) -> i32 {
    std::intrinsics::unchecked_div(a, b)
}
```

I expected to see this happen: MIRI should fail with the following error:
```
error: Undefined Behavior: overflow in signed division (dividing MIN by -1)
```

Instead, this happened: No UB is detected and the program prints `-2147483647`.

### Meta

I ran this using playground. The version reported is:
```
0.1.0 (2023-06-18 2d0aa57)
```

**Comment by celinval** [comment]
I wonder if this is related to https://github.com/rust-lang/rust/pull/112168

**Comment by saethlin** [comment]
I'm confused. `i32::MAX / -1` doesn't overflow, and Miri correctly does not report UB for the unchecked version. `i32::MIN / -1` does overflow, and Miri correctly reports UB for the unchecked version.

```rust
#![feature(core_intrinsics)]

fn main() {
    println!("{:?}", i32::MAX.checked_div(-1));
    let not_ub = unsafe { std::intrinsics::unchecked_div(i32::MAX, -1) };
    println!("{not_ub}");

    println!("{:?}", i32::MIN.checked_div(-1));
    let ub = unsafe { std::intrinsics::unchecked_div(i32::MIN, -1) };
    println!("{ub}");
}
```
Reports UB on the second `unchecked_div`:
```
 --> src/main.rs:9:23
  |
9 |     let ub = unsafe { std::intrinsics::unchecked_div(i32::MIN, -1) };
  |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ overflow in signed division (dividing MIN by -1)
  |
```

**Comment by celinval** [comment]
Shoot, my bad. I did have a typo there, didn't I. Sorry
